### ISSUE TITLE
Does `drift_dev schema dump` has no backward compatibility from 1.1.0?

### ISSUE THREAD

**Issue by Vorkytaka** [simolus3/moor#2743]
I have some JSONs and dart generated files for migrations.

There was no issue before, but currently i'm update to the `drift 2.13.1` with `drift_dev 2.13.2`, update my table and re-run `bump` and `generate`.

And from this point my tests to this new database version is failed because of the schemas.
For example, like this:
```sh
Schema does not match
folder_table:
 columns:
  color:
   Not equal: `` (expected) and `NULL` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)
  order_by_index:
   Not equal: `` (expected) and `NULL` (actual)
```

And, as i see, there is really big difference between JSON from this version and new one.

Maybe, there is an issue with some drift builder's options, but i don't use them.

Can it be fixed somehow from my side?

**Comment by simolus3** [comment]
Thanks for the report. The format has changed, but there is a version tag in the JSON and keeping backwards compatibility is very important.

Does the failure only happen on tests towards the new schema version made after updating to 2.13.1 or on every test? Drift will generate the `NULL` constraint for nullable columns by default, it should also do that when creating the schema used as the expected reference.

**Comment by simolus3** [comment]
Does a schema test like this pass or is this essentially what you're already doing?

```dart
  test('current version matches', () async {
    final database = Database(NativeDatabase.memory());
    await verifier.migrateAndValidate(database, database.schemaVersion);
  });
```

**Comment by Vorkytaka** [comment]
> Thanks for the report. The format has changed, but there is a version tag in the JSON and keeping backwards compatibility is very important.
> 
> Does the failure only happen on tests towards the new schema version made after updating to 2.13.1 or on every test? Drift will generate the `NULL` constraint for nullable columns by default, it should also do that when creating the schema used as the expected reference.

The tests only fails when you test for the new scheme that you get from the new version of Drift.
Like, from previous to new one, or from first one to the latest. Old tests works fine.

Yeah, both json schemas has a version tag. Previous is 1.0 and the new one, that fails, is 1.1.

Unfortunately, i'd reverse changes for now, but if it will help i can reproduce it.
What do you need for that? JSONs? Dart-generated files?

**Comment by simolus3** [comment]
Could you send me the last two schema JSONs? If you don't want to post them publicly you can send them to oss@simonbinder.eu, thanks!

**Comment by Vorkytaka** [comment]
> Could you send me the last two schema JSONs? If you don't want to post them publicly you can send them to [oss@simonbinder.eu](mailto:oss@simonbinder.eu), thanks!

Here is gist with both JSONs and Dart files: https://gist.github.com/Vorkytaka/327805dcc2e2e12be2ed6169b34eb5c4

And tests like this:
```dart
  test('v15 to v16', () async {
    final connection = await verifier.startAt(15);
    final db = MiniDatabase.test(connection.executor);
    await verifier.migrateAndValidate(db, 16);
  });
```

And error like this:
```sh


Schema does not match
todo_table:
 columns:
  date:
   Not equal: `` (expected) and `NULL` (actual)
  time:
   Not equal: `` (expected) and `NULL` (actual)
  completed_date:
   Not equal: `` (expected) and `NULL` (actual)
  folder_id:
   Not equal: `` (expected) and `NULL` (actual)
  note:
   Not equal: `` (expected) and `NULL` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)
  repeat:
   Not equal: `` (expected) and `NULL` (actual)


folder_table:
 columns:
  color:
   Not equal: `` (expected) and `NULL` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)
  order_by_index:
   Not equal: `` (expected) and `NULL` (actual)


subtodo_table:
 columns:
  completed:
   Not equal: `NOT NULL DEFAULT FALSE` (expected) and `NOT NULL DEFAULT 0 CHECK ("completed" IN (0, 1))` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)


notification_table:
 columns:
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)



package:drift_dev/src/services/schema/verifier_impl.dart 151:5  verify
package:drift_dev/src/services/schema/verifier_impl.dart 42:5   VerifierImplementation.migrateAndValidate
```

**Comment by simolus3** [comment]
Thanks for sharing, I can reproduce the issue. Did you migrate these tables from a Dart definition to a definition in `.drift` files between v15 and v16? `todo_table` has `was_declared_in_moor` set to `false` in v15 and to `true` in v16.

**Comment by Vorkytaka** [comment]
> Thanks for sharing, I can reproduce the issue. Did you migrate these tables from a Dart definition to a definition in `.drift` files between v15 and v16? `todo_table` has `was_declared_in_moor` set to `false` in v15 and to `true` in v16.

Yeah, i did.
This was happen a little earlier, not exactly with v16, but i do not create migration for this, and also do not create json-schema.

You think this can be a cause?
Can i fix it somehow?

**Comment by simolus3** [comment]
I think this is the problem, yes. For legacy reasons, drift adds an (entirely meaningless) `NULL` constraint to the definition of nullable columns defined in a Dart file. This doesn't do anything since columns are nullable by default, but it's still expected by the schema verifier. Columns defined in drift files just use the constraints you've written, which doesn't include those `NULL` constraints.

I think the easiest fix would be to add a `NULL` constraints to columns which don't have a `NOT NULL` constraint in the same schema version.

**Comment by Vorkytaka** [comment]
> I think the easiest fix would be to add a `NULL` constraints to columns which don't have a `NOT NULL` constraint in the same schema version.

Not sure if i understand you right.
All my `.drift` nullable columns already contain `NULL`.

**Comment by Vorkytaka** [comment]
> Does a schema test like this pass or is this essentially what you're already doing?
> 
> ```dart
>   test('current version matches', () async {
>     final database = Database(NativeDatabase.memory());
>     await verifier.migrateAndValidate(database, database.schemaVersion);
>   });
> ```

Test looks like this:
```dart
test('v15 to v16', () async {
    final connection = await verifier.startAt(15);
    final db = MiniDatabase.test(connection.executor);
    await verifier.migrateAndValidate(db, 16);
  });
```
and yes, it fails.

**Comment by Vorkytaka** [comment]
> > Does a schema test like this pass or is this essentially what you're already doing?
> > ```dart
> >   test('current version matches', () async {
> >     final database = Database(NativeDatabase.memory());
> >     await verifier.migrateAndValidate(database, database.schemaVersion);
> >   });
> > ```
> 
> Test looks like this:
> 
> ```dart
> test('v15 to v16', () async {
>     final connection = await verifier.startAt(15);
>     final db = MiniDatabase.test(connection.executor);
>     await verifier.migrateAndValidate(db, 16);
>   });
> ```
> 
> and yes, it fails.

With this error:
```
Schema does not match
todo_table:
 columns:
  id:
   Not equal: `NOT NULL PRIMARY KEY DEFAULT (randomblob(16))` (expected) and `NOT NULL DEFAULT (randomblob(16))` (actual)
  date:
   Not equal: `` (expected) and `NULL` (actual)
  time:
   Not equal: `` (expected) and `NULL` (actual)
  completed_date:
   Not equal: `` (expected) and `NULL` (actual)
  folder_id:
   Not equal: `` (expected) and `NULL` (actual)
  note:
   Not equal: `` (expected) and `NULL` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)
  repeat:
   Not equal: `` (expected) and `NULL` (actual)


folder_table:
 columns:
  id:
   Not equal: `NOT NULL PRIMARY KEY DEFAULT (randomblob(16))` (expected) and `NOT NULL DEFAULT (randomblob(16))` (actual)
  color:
   Not equal: `` (expected) and `NULL` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)
  order_by_index:
   Not equal: `` (expected) and `NULL` (actual)


subtodo_table:
 columns:
  id:
   Not equal: `NOT NULL PRIMARY KEY DEFAULT (randomblob(16))` (expected) and `NOT NULL DEFAULT (randomblob(16))` (actual)
  completed:
   Not equal: `NOT NULL DEFAULT FALSE` (expected) and `NOT NULL DEFAULT 0 CHECK ("completed" IN (0, 1))` (actual)
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)


notification_table:
 columns:
  deleted_datetime:
   Not equal: `` (expected) and `NULL` (actual)
```

And this migration:
```
final db15 = v15.DatabaseAtV15(db.executor);
    final db16 = v16.DatabaseAtV16(db.executor);

    await migrator.renameColumn(
      db16.todoTable,
      db15.todoTable.schedule.name,
      db16.todoTable.repeat,
    );
```

And with schemas i send before.

**Comment by Vorkytaka** [comment]
It looks like one of the problems is that when generating via `.tables` files, NULLable columns have empty custom constraints substituted in json, like : `"customConstraints": ""`.

It seems that for this reason we skip [this check](https://github.com/simolus3/drift/blob/7964782c2d82a189b16049972945349f4914bdb1/drift/lib/src/runtime/query_builder/schema/column_impl.dart#L129) and move on.

@simolus3, what do you think?

The problem is that this file generated by drift_dev, so handmade fix will works only till next generation.

**Comment by Vorkytaka** [comment]
But with Primary Key, my bad. 😥
I didn't add Primary Key in the old dart-table, but I added it to the .drift, so I got an error.

**Comment by Vorkytaka** [comment]
> It looks like one of the problems is that when generating via `.tables` files, NULLable columns have empty custom constraints substituted in json, like : `"customConstraints": ""`.
> 
> It seems that for this reason we skip [this check](https://github.com/simolus3/drift/blob/7964782c2d82a189b16049972945349f4914bdb1/drift/lib/src/runtime/query_builder/schema/column_impl.dart#L129) and move on.
> 
> @simolus3, what do you think?

Create a small PR for that: https://github.com/simolus3/drift/pull/2786

**Comment by simolus3** [comment]
> All my `.drift` nullable columns already contain `NULL`.

Ouch, it looks like the generator is dropping them. I have fixed that in 08ceb939bc2869c9c31b8f4c18828278f68d459d.

But `NULL` in a column definition has no meaning since columns are nullable by default. Of course, it's very unfortunate that a transition from Dart tables to drift tables is infeasible for nullable columns because `NULL`s are dropped in drift files. This will be fixed in drift 2.15. Until then, I'm afraid you'd have to bump the schema version and run an empty `alterTable(TableMigration(table))` so that drift removes the `NULL` constraint from the schema.

**Comment by Vorkytaka** [comment]
> > It looks like one of the problems is that when generating via `.tables` files, NULLable columns have empty custom constraints substituted in json, like : `"customConstraints": ""`.
> > It seems that for this reason we skip [this check](https://github.com/simolus3/drift/blob/7964782c2d82a189b16049972945349f4914bdb1/drift/lib/src/runtime/query_builder/schema/column_impl.dart#L129) and move on.
> > @simolus3, what do you think?
> 
> Create a small PR for that: #2786

Ok, no, it's looks like you planned this behavior?
Some tests with custom constraints are failed.

**Comment by Vorkytaka** [comment]
> > All my `.drift` nullable columns already contain `NULL`.
> 
> Ouch, it looks like the generator is dropping them. I have fixed that in [08ceb93](https://github.com/simolus3/drift/commit/08ceb939bc2869c9c31b8f4c18828278f68d459d).
> 
> But `NULL` in a column definition has no meaning since columns are nullable by default. Of course, it's very unfortunate that a transition from Dart tables to drift tables is infeasible for nullable columns because `NULL`s are dropped in drift files. This will be fixed in drift 2.15. Until then, I'm afraid you'd have to bump the schema version and run an empty `alterTable(TableMigration(table))` so that drift removes the `NULL` constraint from the schema.

Ok, thanks!
Then I'll close the issue.
